drf/1/myProject/books/views.py

```diff
@@ -0,0 +1,142 @@
+from rest_framework.decorators import api_view
+from .serializers import BookSerializer
+from django.http import HttpResponse, JsonResponse
+from rest_framework.request import Request
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from rest_framework import authentication, permissions
+from django.shortcuts import get_object_or_404
+from django.views.decorators.csrf import csrf_exempt
+
+
+# Q1:
+def get_all_books(request):
+    if request.method == "GET":
+        try:
+            queryset = Book.objects.all()
+            return HttpResponse(queryset, status=200)
+
+        except:
+            return HttpResponse("cant get books", status=404)
+
+
+def get_jk_rowling(request):
+    if request.method == "GET":
+        try:
+            queryset = Book.objects.get(author="JK Rowling")
+            return HttpResponse(queryset, status=200)
+
+        except:
+            return HttpResponse("cant get books", status=404)
+
+
+# Q2:
+# .data returns the parsed content of the request.
+
+
+@api_view(["POST"])
+def create_book(request):
+    request_data = request.data
+    serializer = BookSerializer(data=request_data)
+    if serializer.is_valid():
+        serializer.save()
+        return Response(request_data, status=status.HTTP_201_CREATED)
+
+    else:
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+@api_view(["GET"])
+def retrieve_book(request):
+    name = request.query_params.get("title")
+    book = get_object_or_404(Book, title=name)
+    book_data = BookSerializer(book).data
+    return Response(book_data, status=200)
+
+
+@api_view(["POST"])
+def update_book(request):
+    data = request.data
+    book = get_object_or_404(Book, title = data["title"])
+    book_serializer = BookSerializer(book, data)
+    if book_serializer.is_valid():
+        book_serializer.update(book, data)
+        return Response("Book was updated!", status=200)
+
+    else:
+        return Response(book_serializer.errors, status=404)
+
+
+@api_view(["DELETE"])
+def delete_book(request, name):
+    print(name)
+    book = get_object_or_404(Book, title = name)
+    book.delete()
+    return Response("deleted", status=200)
+
+
+# Q3:
+# query params is like request.GET but better. It is a dictionary-like object that allows us to access the query parameters sent with the URL as part of an HTTP request.
+
+# Q4:
+# the difference between Response() and HttpResponse() is that unlike HttpResponse objects, you dont instantiate Response objects with rendered content.
+
+# Q5:
+# class-based-views : drf provides an APIView class which subclasses Django view class.
+# Request instances will be passed to the handler and not HttpRequest instances.
+# the handler will return Response and not HttpResponse. 
+
+# function-based-views : provides a set of simple decorators that wrap your function based views to ensure they receive an instance of Request.
+# and allows them to return a Response.
+
```

Comment:
you didn't answer fully- when do we want to use each type. also, you mixed the 2 types in your views.